File: src/main/java/jp/tsubakicraft/mongocrud/controller/BrandController.java
package jp.tsubakicraft.mongocrud.controller;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import jp.tsubakicraft.mongocrud.model.Brand;
import jp.tsubakicraft.mongocrud.service.BrandRepository;

@RestController
public class BrandController {

	@Autowired
	private BrandRepository repo;

	@RequestMapping(value = "/api/brands/listAll", method = RequestMethod.GET)
	public List<Brand> listAll() {
		Sort sort = new Sort(Sort.Direction.ASC, "name");
		return repo.findAll(sort);
	}

	@RequestMapping(value = "/api/brands", method = RequestMethod.GET)
	public Page<?> listBrands(@RequestParam(value = "page", required = true) int page,
			@RequestParam(value = "limit", required = true) int limit,
			@RequestParam(value = "sortColumn", required = true) String column,
			@RequestParam(value = "sortDir", required = true) String dir) {
		Sort sort = new Sort(
				new Sort.Order("asc".equalsIgnoreCase(dir) ? Sort.Direction.ASC : Sort.Direction.DESC, column));
		Pageable pageRequest = new PageRequest(page, limit, sort);
		Page<Brand> p = repo.findAll(pageRequest);
		return p;
	}

	@RequestMapping(value = "/api/brands", method = RequestMethod.PUT)
	public Brand updateBrand(@RequestBody Brand brand) {
		Brand b = repo.findOne(brand.id);
		if (b != null) {
			b.name = brand.name;
			repo.save(b);
		}
		return b;
	}

	@RequestMapping(value = "/api/brands", method = RequestMethod.DELETE)
	public Brand deleteBrand(@RequestBody Brand brand) {
		repo.delete(brand.id);
		return brand;
	}

	@RequestMapping(value = "/api/brands", method = RequestMethod.POST)
	public Brand createBrand(@RequestBody Brand brand) {
		Brand b = new Brand();
		b.name = brand.name;
		repo.save(b);
		return b;
	}
}

File: src/main/resources/static/scripts/brand_controller.js
app.controller("brand_controller", function($scope, $http, $location, $q, ngDialog) {
	
	$scope.page = 0;
	$scope.limit = 5;
	$scope.brands = null;
	$scope.currentPage = 1;
	$scope.totalItems = 0;
	$scope.brandId = "";
    $scope.show = true;
    $scope.hide = true;
	$scope.brand = {};
	$scope.sortColumn = "name";
	$scope.sortDir = "ASC";
	
	$scope.listBrands = function() {
		$http.get("/api/brands", {params: {page: $scope.page, limit: $scope.limit, sortColumn: $scope.sortColumn, sortDir: $scope.sortDir}}).then(function(response) {
			$scope.brands = response.data;
			$scope.currentPage = $scope.brands.number+1;
			$scope.totalItems = $scope.brands.totalElements;
		}, function(error) {
			$scope.error = error;
		});
	};
	
	$scope.createBrand = function() {
		if(!$scope.validateForm()) {
			$http.post("/api/brands", $scope.brand).then(function(response) {
		        $scope.show = true;
		        $scope.hide = true;
		        $scope.hideObj = false;
		        $scope.showObj = false;
		        $scope.brandId = "";
		    	$location.path("/brand");
			}, function(error) {
				$scope.error = error;
			});
		}
	};
	
	$scope.editBrand = function(brandId) {
        $scope.show = false;
        $scope.hide = false;
        $scope.hideObj = true;
        $scope.showObj = true;
        $scope.brandId = brandId;
	};
	
	$scope.deleteBrand = function(brand) {
		$scope.brand = brand;
		$http.get("/api/models/countBrandModel", {params:{brandId: brand.id}}).then(function(response) {
			var count = response.data;
			if(count == 0) {
				ngDialog.openConfirm({
					scope: $scope,
					template: 'views/brandDeleteConfirm.html',
					className: 'ngdialog-theme-default',
					showClose: false
				}).then(function(value) {
					$http.delete("/api/brands", {data: brand, headers: {'Content-Type':'application/json; charset=UTF-8'}}).then(function(response) {
						$scope.listBrands();
					}, function(error) {
						$scope.error = error;
					});
				}, function(value) {
				});
			} else {
				ngDialog.open({
					template: '<p>Brand <strong>' + brand.name + '</strong> cannot be deleted. There are ' + count + ' models exists.</p>\
					<div class="ngdialog-buttons">\
					<button class="ngdialog-button ngdialog-button-secondary" ng-click="closeThisDialog(0)">OK</button></div>',
					plain: true,
					showClose: false
				});
			}
		}, function(error) {
			$scope.error = error;
		});
	};
	
	$scope.updateBrand = function(brand) {
		$http.put("/api/brands", brand).then(function(response) {
	        $scope.show = true;
	        $scope.hide = true;
	        $scope.hideObj = false;
	        $scope.showObj = false;
	        $scope.brandId = "";
	    	$location.path("/brand");
		}, function(error) {
			$scope.error = error;
		});
	};
	
	$scope.undoEdit = function() {
        $scope.show = true;
        $scope.hide = true;
        $scope.hideObj = false;
        $scope.showObj = false;
        $scope.brandId = "";
	};
	
	$scope.setPage = function() {
	    $scope.currentPage = pageNo;
	};
	
	$scope.pageChanged = function() {
	    $scope.page = $scope.currentPage - 1;
	    $scope.listBrands();
	};
	
	$scope.linkTo = function(path) {
		$location.path(path);
	};
	
	$scope.sort = function(column) {
		if($scope.sortColumn == column) {
			$scope.sortDir = $scope.sortDir == "ASC" ? "DESC" : "ASC";
		} else {
			$scope.sortDir = "ASC";
			$scope.sortColumn = column;
		}
		$scope.listBrands();
	};
	
	$scope.validateForm = function() {
		$scope.validationMessages = [];
		if($scope.brand.name == null || $scope.brand.name == null) {
			$scope.validationMessages.push("Name is required.");
		}
		$scope.hasErrors =  $scope.validationMessages.length > 0;
		return $scope.hasErrors;
	};

	$scope.listBrands();
});
